### PR TITLE
Implement movable instances

### PR DESCRIPTION
**Pull request by MrTroble**
Changed instance read write to a new bundle system

### CODE REVIEW

File: ShaderTool/Command/Map.cs [hunk 1]
```diff
@@ -486,30 +497,32 @@ class Map {
...
-                // Find the material ID from the material name
-                byte id = (byte)Array.IndexOf(mapData.materialNames, actorData.materialName);
+                // List is always a multiple of 4
+                int actorTransformID = transformList.Count / 4;
+                StreamWrite(actorTransformID);
 
                 // Just take the last material in case none was supplied
                 // We can avoid excessive error handling that way
-                StreamWrite(id);
+                int materialindex = Array.IndexOf(mapData.materialNames, actorData.materialName);
```
Comment (IceDynamix) [inline ShaderTool/Command/Map.cs:514]: why was this changed to an int? isnt the limit still 256?

File: ShaderTool/Command/Map.cs [hunk 2]
```diff
@@ -531,41 +544,48 @@ class Map {
...
+                    data.animationId = 0;
+                    instanceList.Add(data);
 
                     transformList.AddRange(instanceMatrix);
                 }
 
             }
 
-            StreamWrite(0xFFFFFFFF);
+            StreamWrite(instanceList.Count);
```
Comment (IceDynamix) [inline ShaderTool/Command/Map.cs:564]: you already wrote the instance count in line 524 though? why was the max int removed?

File: ShaderTool/Command/Material.cs [hunk 3]
```diff
@@ -17,7 +17,7 @@ class MaterialData {
     class Material {
 
         public static string MaterialPath = Path.Combine(Program.ResourcesFolder, @"Materials.json");
-        public const uint MATERIAL_SIZE = 24; // Color is 4 * 4 bytes, diffuse texture ID and animation is 4 bytes * 2 which equals 24 bytes
+        public const uint MATERIAL_SIZE = 20; // Color is 4 * 4 bytes, diffuse texture ID and animation is 4 bytes * 2 which equals 24 bytes
```
Comment (IceDynamix) [inline ShaderTool/Command/Material.cs:20]: comment should be changed to reflect the current value

File: ShaderTool/Command/Map.cs [hunk 4]
```diff
@@ -474,7 +479,13 @@ class Map {
 
         }
 
-        private static int AddActorsToResource(ref List<float> transformList) {
+        struct _TransformData{
```
Comment (IceDynamix) [inline ShaderTool/Command/Map.cs:482]: underscore?

File: TGEngine/gamecontent/Actor.hpp [hunk 5]
```diff
@@ -8,28 +8,49 @@
 namespace tge::gmc {
 
 	/*
-     * This is struct holds all the properties that
+     * This struct holds all the transform information
+	 * used within a drawcall by this actor
+     *
+     * <ul>
+     * <li><strong class='atr'>matrix</strong> are the tranform information</li>
```
Comment (IceDynamix) [inline TGEngine/gamecontent/Actor.hpp:15]:
```suggestion
     * <li><strong class='atr'>matrix</strong> is the transform information</li>
```

File: TGEngine/gamecontent/Actor.hpp [hunk 6]
```diff
@@ -8,28 +8,49 @@
...
+     */
+	struct ActorTransform {
+		glm::fmat4 matrix;
+		uint32_t  animationIndex;
+		uint32_t  transformIndex;
+	};
+
+	/*
+     * This struct holds all the properties that
 	 * do not directly influence the mesh it self
```
Comment (IceDynamix) [inline TGEngine/gamecontent/Actor.hpp:33]:
```suggestion
	 * do not directly influence the mesh itself
```

File: TGEngine/gamecontent/Actor.hpp [hunk 7]
```diff
@@ -8,28 +8,49 @@
...
      *
      * <ul>
-     * <li><strong class='atr'>localTransform</strong> is the local tranform of the actor that will be computed at runtime</li>
+     * <li><strong class='atr'>transform</strong> are the tranform information</li>
```
Comment (IceDynamix) [inline TGEngine/gamecontent/Actor.hpp:36]:
```suggestion
     * <li><strong class='atr'>transform</strong> is the transform information</li>
```

File: TGEngine/gamecontent/Actor.hpp [hunk 8]
```diff
@@ -8,28 +8,49 @@
...
      * <li><strong class='atr'>material</strong> is the index of the material used to draw this mesh</li>
      * <li><strong class='atr'>layer</strong> is the layer the actor is drawn on (e.g. 0 for normal world)</li></ul>
      *
      * <h4>Valid usage</h4>
      * <ul>
-     * <li><strong class='atr'>localTransform</strong> must be a non-zero matrix otherwise this actor will be ignored</li>
+     * <li><strong class='atr'>transform</strong> must be a valid ActorTransform struct</li>
      * <li><strong class='atr'>material</strong> must be a valid material id in the material list</li>
      * <li><strong class='atr'>layer</strong> must be a valid layer id otherwise this actor will be ingnored</li></ul>
```
Comment (IceDynamix) [inline TGEngine/gamecontent/Actor.hpp:44]:
```suggestion
     * <li><strong class='atr'>layer</strong> must be a valid layer id, otherwise this actor will be ingnored</li></ul>
```

File: TGEngine/gamecontent/Actor.hpp [hunk 9]
```diff
@@ -8,28 +8,49 @@
...
-		uint32_t  layer;
+		ActorTransform transform;
+		uint32_t       material;
+		uint32_t       layer;
 	};
 
 	/*
-     * This is struct holds all the information
+     * This struct holds all the information
      * needed to draw an actor such as offsets in the buffer
```
Comment (IceDynamix) [inline TGEngine/gamecontent/Actor.hpp:54]:
```suggestion
     * needed to draw an actor, such as offsets in the buffer
```

File: TGEngine/pipeline/Descriptors.cpp [hunk 10]
```diff
@@ -68,10 +68,10 @@ void initDescriptors() {
 	VkPushConstantRange pushConstantRanges[2];
 	pushConstantRanges[0].stageFlags = VK_SHADER_STAGE_VERTEX_BIT;
 	pushConstantRanges[0].offset = 0;
-	pushConstantRanges[0].size = 68; // mat4 + animationIndex
+	pushConstantRanges[0].size = 72; // mat4 + animationIndex + tranformIndex
```
Comment (IceDynamix) [inline TGEngine/pipeline/Descriptors.cpp:71]: you do like your tranforms, dont you
```suggestion
	pushConstantRanges[0].size = 72; // mat4 + animationIndex + transformIndex
```

File: ShaderTool/Command/Map.cs (hunk 2 above)
Comment (MrTroble) [reply ShaderTool/Command/Map.cs:564]: To preallocate memory in the actor read function.

File: ShaderTool/Command/Map.cs (hunk 1 above)
Comment (MrTroble) [reply ShaderTool/Command/Map.cs:514]: This has been changed primarily for alignment purposes.

File: ShaderTool/Command/Material.cs [hunk 11]
```diff
@@ -17,7 +17,7 @@ class MaterialData {
     class Material {
 
         public static string MaterialPath = Path.Combine(Program.ResourcesFolder, @"Materials.json");
-        public const uint MATERIAL_SIZE = 24; // Color is 4 * 4 bytes, diffuse texture ID and animation is 4 bytes * 2 which equals 24 bytes
+        public const uint MATERIAL_SIZE = 20; // Color is 4 * 4 bytes, diffuse texture ID is 4 bytes which equals 24 bytes
```
Comment (IceDynamix) [inline ShaderTool/Command/Material.cs:20]: 4*4+4=24
```suggestion
        public const uint MATERIAL_SIZE = 20; // Color is 4 * 4 bytes, diffuse texture ID is 4 bytes which equals 20 bytes
```
